File: ts/project/.sdk/tm/java/utility/TransformRequest.java
package JAVAPACKAGE.utility;

import java.util.LinkedHashMap;
import java.util.Map;

import JAVAPACKAGE.core.Context;
import JAVAPACKAGE.core.Helpers;
import JAVAPACKAGE.utility.struct.Struct;

final class TransformRequest {

  private TransformRequest() {}

  static Object transformRequest(Context ctx) {
    if (ctx.spec != null) {
      ctx.spec.step = "reqform";
    }

    Map<String, Object> transform =
        Helpers.toMapAny(Struct.getprop(ctx.point, "transform"));
    if (transform == null) {
      return stripAction(ctx.reqdata);
    }

    Object reqform = Struct.getprop(transform, "req", null);
    if (reqform == null) {
      return stripAction(ctx.reqdata);
    }

    Map<String, Object> data = new LinkedHashMap<>();
    data.put("reqdata", ctx.reqdata);

    return stripAction(Struct.transform(data, reqform));
  }

  // `$action` selects the point (see MakePoint); it is never an API field, so
  // the body is a copy without it. The caller's map is left untouched.
  private static Object stripAction(Object reqdata) {
    if (!(reqdata instanceof Map) || !((Map<?, ?>) reqdata).containsKey("$action")) {
      return reqdata;
    }
    Map<String, Object> body = new LinkedHashMap<>();
    for (Map.Entry<?, ?> e : ((Map<?, ?>) reqdata).entrySet()) {
      if (!"$action".equals(e.getKey())) {
        body.put(String.valueOf(e.getKey()), e.getValue());
      }
    }
    return body;
  }
}

File: ts/project/.sdk/tm/rust/utility/transform_request.rs
use std::rc::Rc;

use crate::core::context::Context;
use crate::core::helpers::{getp, jo, setp, to_map};
use crate::utility::voxgigstruct as vs;
use crate::utility::voxgigstruct::Value;

// `$action` selects the point (see make_point_util); it is never an API
// field, so the body is a copy without it. The caller's map is left untouched.
fn strip_action(reqdata: Value) -> Value {
    let has = match &reqdata {
        Value::Map(src) => src.borrow().contains_key("$action"),
        _ => false,
    };
    if !has {
        return reqdata;
    }
    let body = Value::empty_map();
    if let Value::Map(src) = &reqdata {
        let entries: Vec<(String, Value)> = src
            .borrow()
            .iter()
            .map(|(k, v)| (k.clone(), v.clone()))
            .collect();
        for (k, v) in entries {
            if "$action" != k {
                setp(&body, &k, v);
            }
        }
    }
    body
}

pub fn transform_request_util(ctx: &Rc<Context>) -> Value {
    let spec = ctx.spec.borrow().clone();
    let point = ctx.point.borrow().clone();

    if let Some(sp) = &spec {
        sp.borrow_mut().step = "reqform".to_string();
    }

    let transform = to_map(&getp(&point, "transform"));
    if transform.is_noval() {
        return strip_action(ctx.reqdata.borrow().clone());
    }

    let reqform = getp(&transform, "req");
    if reqform.is_noval() || reqform.is_null() {
        return strip_action(ctx.reqdata.borrow().clone());
    }

    let store = jo(vec![("reqdata", ctx.reqdata.borrow().clone())]);
    strip_action(match vs::transform(&store, &reqform, None) {
        Ok(reqdata) => reqdata,
        Err(_) => ctx.reqdata.borrow().clone(),
    })
}
